doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+# Dynamic resource sharing for sources and sinks
+
+## Summary
+
+In Materialize today, each source and sink instance is independent. For example:
+
+  * Each instance of a Kafka source creates its own Kafka consumer.
+  * Each instance of a file source creates its own file handle and filesystem
+    event listener.
+  * Each instance of a `TAIL` sink performs its own consolidation of the input
+    stream.
+
+The result is a system that can correctly backfill data when a source is
+instantiated multiple times, but is inefficient in the steady state. In an ideal
+system, once two instances of a source or sink had caught up to the same point,
+they would share resources. To continue the example from above:
+
+  * Multiple instances of a Kafka source would share the underlying Kafka
```

Comment:
The ticket for specifically sharing multiple instances of a Kafka source is #3791